### PR TITLE
[AIRFLOW-184] Add mark success to CLI

### PR DESCRIPTION
**Pull request by jgao54**
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- _(replace with a link to AIRFLOW-184)_

Reminders for contributors:
- Your PR's title must reference an issue on [Airflow's JIRA](https://issues.apache.org/jira/browse/AIRFLOW/). For example, a PR called "[AIRFLOW-1] My Amazing PR" would close JIRA issue #1. Please open a new issue if required!
- You must add an [Apache License header](http://www.apache.org/legal/src-headers.html) to all new files.
- Please squash your commits when possible and follow the [7 rules of good Git commits](http://chris.beams.io/posts/git-commit/#seven-rules).

### CODE REVIEW

**Comment by criccomini** [comment]
Looks like tests are unhappy:

```
TypeError: mark_success() got an unexpected keyword argument 'confirm_propt'
```

Typo? :)

Also saw this error, which looks a little more serious:

```
assert "Wait a minute" in response.data.decode('utf-8')
AssertionError
```

File: airflow/bin/cli.py [hunk 1]
```diff
@@ -411,6 +411,22 @@ def clear(args):
...
+    if args.task_regex:
+        dag = dag.sub_dag(
+            task_regex=args.task_regex,
+            include_downstream=args.downstream,
+            include_upstream=args.upstream)
+
+    dag.mark_success(
+        start_date=args.start_date,
+        end_date=args.end_date,
+        confirm_propt=not args.no_confirm,
```
Comment (criccomini) [inline airflow/bin/cli.py]: Typo? prompt

File: airflow/models.py [hunk 2]
```diff
@@ -2813,6 +2813,64 @@ def clear(
...
+                     start_date=None,
+                     end_date=None,
+                     confirm_prompt=False,
+                     include_subdags=True,
+                     dry_run=False,
+                     session=None):
+        """
+        Mark success a list of task instances associated with the current dag
+        for a specific date range.
+        """
```
Comment (criccomini) [inline airflow/models.py]: Nit: pydoc param stuff would be nice here (:type: and :param:)

File: airflow/www/views.py [hunk 3]
```diff
@@ -1117,100 +1116,35 @@ def success(self):
...
-            if recursive:
-                recurse_tasks(relatives, task_ids, dag_ids, task_id_to_dag)
-        if upstream:
-            relatives = task.get_flat_relatives(upstream=False)
-            task_ids += [t.task_id for t in relatives]
-            if recursive:
-                recurse_tasks(relatives, task_ids, dag_ids, task_id_to_dag)
-        TI = models.TaskInstance
+        dag = dag.sub_dag(
+            task_regex=r"^{0}$".format(task_id),
```
Comment (criccomini) [inline airflow/www/views.py]: Not sure--does the task_id need to be escaped here? Wondering if task_id of "foo.bar" would catch both "foo.bar" and "foo_bar"?

File: airflow/www/views.py [hunk 4]
```diff
@@ -1117,100 +1116,35 @@ def success(self):
...
+            include_downstream=downstream,
+            include_upstream=upstream)
 
-        if dag.schedule_interval == '@once':
```
Comment (criccomini) [inline airflow/www/views.py]: Does new code handle this properly? Might be worth a test.

File: airflow/www/views.py [hunk 5]
```diff
@@ -1117,100 +1116,35 @@ def success(self):
...
-        tasks = list(product(task_ids, dates))
-        tis_to_create = list(
-            set(tasks) -
-            set([(ti.task_id, ti.execution_date) for ti in tis]))
-
-        tis_all_altered = list(chain(
-            [(ti.task_id, ti.execution_date) for ti in tis_to_change],
-            tis_to_create))
-
-        if len(tis_all_altered) > MAX_PERIODS:
```
Comment (criccomini) [inline airflow/www/views.py:1054]: (I'm being lazy) is MAX_PERIODS used anywhere else? If not, should delete.

File: airflow/www/views.py [hunk 6]
```diff
@@ -1117,100 +1116,35 @@ def success(self):
...
-            TI.dag_id.in_(dag_ids),
-            TI.execution_date.in_(dates),
-            TI.task_id.in_(task_ids)).all()
-        tis_to_change = session.query(TI).filter(
-            TI.dag_id.in_(dag_ids),
-            TI.execution_date.in_(dates),
-            TI.task_id.in_(task_ids),
-            TI.state != State.SUCCESS).all()
-        tasks = list(product(task_ids, dates))
-        tis_to_create = list(
```
Comment (criccomini) [inline airflow/www/views.py:1046]: @bolkedebruin I think we need some clarity from you on this one. @mistercrunch said this code is to create a task for instances that haven't yet been created, so that they can be marked as success.

But, I think this pr (https://github.com/apache/incubator-airflow/pull/1506) might invalidate this logic. Question is: do we need to create tasks anymore, or does searching for state = None and state != success suffice?

**Comment by bolkedebruin** [comment]
I think the overall approach needs to be updated. Ie. you do not mark a "dag" as a success but your mark a "DagRun" and/org its underlying tasksinstances as a success. As such I dont like "dag.mark_success".

Creating TIs outside a dagrun is becoming a big no-no and I would say it is even worse to do so in a "view" component, besides it is indeed not required anymore since #1506 as tasks are eagerly created. 

All in all, I really like the functionality but please make it DagRun based.

File: airflow/www/views.py (hunk 6 above)
Comment (bolkedebruin) [inline airflow/www/views.py:1046]: In principle you should create tasks anymore. We verify the integrity of DagRuns now and re-create tasks that do not exist in a DagRun. DagRun also exposes "get_task_instances" that is the proper way to get tasks.

Comment (criccomini) [inline airflow/www/views.py:1046]: Just to confirm, sounds like it shouldn't be an issue that this patch no longer inserts the task instances, then, correct?

**Comment by criccomini** [comment]
Agree with @bolkedebruin's comments about DAG run vs dag.

File: airflow/www/views.py (hunk 6 above)
Comment (bolkedebruin) [inline airflow/www/views.py:1046]: Correct

File: airflow/www/views.py (hunk 3 above)
Comment (jgao54) [inline airflow/www/views.py]: Good catch!

File: airflow/www/views.py (hunk 4 above)
Comment (jgao54) [inline airflow/www/views.py]: This is out of scope now since it's dagrun-based instead of dag-based. For sanity check, I ran a manual test and can confirm.

File: airflow/bin/cli.py [hunk 7]
```diff
@@ -424,6 +424,50 @@ def clear(args):
         confirm_prompt=not args.no_confirm,
         include_subdags=not args.exclude_subdags)
 
+def mark_success(args):
+    dag = get_dag(args)
+    dagruns = dag.get_dagruns(args.start_date, args.end_date)
+    # todo: run.dag is transient, needs to be set
```
Comment (jgao54) [inline airflow/bin/cli.py]: There seem to be a few places in the code where this is manually set so I'm leaving it out of the scope of this PR.

File: airflow/models.py [hunk 8]
```diff
@@ -3444,7 +3458,7 @@ def find(dag_id=None, run_id=None, execution_date=None,
         return dr
 
     @provide_session
-    def get_task_instances(self, state=None, session=None):
+    def get_task_instances(self, state=None, to_list=True, session=None):
```
Comment (criccomini) [inline airflow/models.py]: Nit: this is backwards incompatible, and could lead to issues if someone was using unnamed params and providing a session. Worth a quick glance, but I'm OK with this change.

Comment (criccomini) [inline airflow/models.py]: Actually, the only place that I see this used is in jobs.py (twice), and in tests. Would prefer removing the to_list param, and just returning tis. Then updating code that makes the call to use .all() on the return object.

Ack that this is backwards compatible, but seems very limited usage.

File: airflow/bin/cli.py (hunk 7 above)
Comment (criccomini) [inline airflow/bin/cli.py]: TODO still required?

File: airflow/bin/cli.py [hunk 9]
```diff
@@ -424,6 +424,50 @@ def clear(args):
...
+    for run in dagruns:
```
Comment (criccomini) [inline airflow/bin/cli.py]: @bolkedebruin this logic is kinda funky. Any thoughts on this?

File: airflow/bin/cli.py [hunk 10]
```diff
@@ -424,6 +424,50 @@ def clear(args):
...
+            include_upstream=args.upstream,
+            include_subdags=not args.exclude_subdags,
+            dry_run=True)
+
+    if len(tis) == 0:
+        print("No task instances to mark as successful")
+        return
+
+    if len(tis) > 1000:
+        print("Too man tasks (>1000)")
```
Comment (criccomini) [inline airflow/bin/cli.py]: Typo

File: airflow/bin/cli.py [hunk 11]
```diff
@@ -424,6 +424,50 @@ def clear(args):
...
+def mark_success(args):
+    dag = get_dag(args)
+    dagruns = dag.get_dagruns(args.start_date, args.end_date)
+    # todo: run.dag is transient, needs to be set
+    for run in dagruns:
+        run.dag = dag
+
+    tis = []
+    for run in dagruns:
+        tis += run.mark_success(
```
Comment (criccomini) [inline airflow/bin/cli.py]: It's kind of annoying to have to do this twice.. once with dry_run and once with out.. Just complaining.

File: airflow/models.py [hunk 12]
```diff
@@ -3092,6 +3092,20 @@ def create_dagrun(self,
...
+        """
+        Return a list of dag runs from this dag with execution date between
+        start_date and end_date
+        """
+        DR = DagRun
+        dr = session.query(DR).filter(DR.dag_id == self.dag_id)
+        if start_date:
+            dr = dr.filter(DR.execution_date >= start_date)
+        if end_date:
+            dr = dr.filter(DR.execution_date <= end_date)
```
Comment (criccomini) [inline airflow/models.py:3379]: It annoys me that this is inclusive on end_date. If other CLI params behave this way, then OK. If not, would prefer exclusive end_date.

**Comment by criccomini** [comment]
Took another pass and left some comments.

Make sure to take a look at the failing test. Looks like a legit failure after your changes.

Also, we should block on @bolkedebruin for this one, I think, since it's touching DagRun, and he's refactoring stuff there.

**Comment by artwr** [comment]
Just curious to see if you still intend to refactor this, because I think there is a definite use case for this to be enabled. But maybe this can wait for a REST API.

**Comment by jgao54** [comment]
@artwr  @criccomini  @bolkedebruin yeah, definitely still intend to refactor this. The current mark success in the UI has a few bugs, (using mark_success in conjunction with past, upstream, or downstream flag throws the error page in some cases).  This pr will address those issues as well. I'm ok with waiting for REST API if it's coming up soon, otherwise I think it may be more beneficial to have this merged (once I clean it up =P ).

File: airflow/bin/cli.py (hunk 11 above)
Comment (jgao54) [inline airflow/bin/cli.py]: I agree it's a little annoying, plus it creates an illusion of what is selected the first pass is definitely what is marked success (i.e. rare corner case which that isn't true is if 'future' is selected, and a TI that instantiate after the first pass will get marked success without being listed).

File: airflow/models.py (hunk 12 above)
Comment (jgao54) [inline airflow/models.py:3379]: Inclusive end_date has been used so far, but use case for cli is only airflow/clear, so seems ok to change.

**Comment by r39132** [comment]
I think a mark success feature is a good one.  I have been looking for one and don't like running a backfill to achieve it. There is one issue I noticed on the current 1.7.1.3 release.

I was 3 days behind on an hourly run. I was hoping a single `backfill -m` would suffice to mark everything success till current. However, since DAG Runs/Task Instances are generated in little groups by scheduler cycles, that was not the case.... I had to continuously run mark success.. for an hour.

**Comment by criccomini** [comment]
> I'm ok with waiting for REST API if it's coming up soon, otherwise I think it may be more beneficial to have this merged (once I clean it up =P ).

I think it's worth getting this PR done. The REST API is likely a long ways away.

File: airflow/models.py (hunk 12 above)
Comment (jgao54) [inline airflow/models.py:3379]: decided to keep it as is for now since 'airflow clear' cli is probably used in prod.

**Comment by jgao54** [comment]
@bolkedebruin I noticed that subdag's TIs are created independently rather than via DagRun (there are currently no dagrun being created for subdags). Was wondering if this is intentional? imho it would make sense to create dagruns for subdags as well if we want to support mark success / clear a subdag via dagrun.

**Comment by bolkedebruin** [comment]
@jgao54 that is indeed so, but old style. I have a PR outstanding that does add dag runs to backfills (that is how subdags run) but it needs reviewing/testing

**Comment by jgao54** [comment]
@bolkedebruin ah I see it! In that case I'll hold off this PR until yours is merged. I will need to make a bit of changes accordingly, but it will be cleaner.

**Comment by bolkedebruin** [comment]
If you can please review the other pr and +1 it. Thanks!

**Comment by criccomini** [comment]
@bolkedebruin what's the PR?

**Comment by jgao54** [comment]
@criccomini  https://github.com/apache/incubator-airflow/pull/1667

**Comment by jgao54** [comment]
@criccomini  @bolkedebruin  @jlowin  Updated this PR to reflect the changes from #1667

**Comment by criccomini** [comment]
cc @aoen

**Comment by criccomini** [comment]
Seems like all tests are failing?

**Comment by jgao54** [comment]
@criccomini thanks! fixed

**Comment by r39132** [comment]
@jgao54 Resurrecting this.. please rebase and prepare for review or request to close.

**Comment by r39132** [comment]
@jgao54 Do you still want to merge this? If not, I will close for inactivity.

**Comment by r39132** [comment]
@jgao54 closing for inactivity. Please reopen when ready to submit.
